sdlib/d/gc/slab.d

```diff
@@ -23,6 +25,218 @@ struct SlabAllocGeometry {
 	}
 }
 
+struct SlabAllocInfo {
+private:
+	SlabAllocGeometry sg;
+	Extent* e;
+	bool scSupportsMetaData = false;
+	bool metaDataEnabled = false;
+
+public:
+	@property
+	auto allowsMetaData() {
+		return scSupportsMetaData;
+	}
+
+	@property
+	auto hasMetaData() {
+		return metaDataEnabled;
+	}
+
+	this(PageDescriptor pd, const void* ptr) {
+		assert(pd.isSlab(), "Expected a slab!");
+
+		e = pd.extent;
+		sg = SlabAllocGeometry(pd, ptr);
+		scSupportsMetaData = sizeClassSupportsMetadata(pd.sizeClass);
+		metaDataEnabled = scSupportsMetaData && e.hasMetaData(sg.index);
+	}
+
+	@property
+	auto address() {
+		return sg.address;
+	}
+
+	@property
+	size_t slotCapacity() {
+		return sg.size;
+	}
+
+	@property
+	size_t usedCapacity() {
+		return sg.size - freeSpace;
+	}
+
+	bool setUsedCapacity(size_t size) {
+		if (!scSupportsMetaData || (size > slotCapacity)) {
+			return false;
+		}
+
+		setFreeSpace(sg.size - size);
+		return true;
+	}
+
+	@property
+	size_t freeSpace() {
+		return metaDataEnabled ? readPackedFreeSpace(freeSpacePtr) : 0;
+	}
+
+	void setFreeSpace(size_t size) {
+		assert(scSupportsMetaData, "size class not supports slab metadata!");
+		assert(size <= sg.size, "size exceeds alloc size!");
+
+		if (size == 0) {
+			disableMetaData();
+			return;
+		}
+
+		writePackedFreeSpace(freeSpacePtr, size & ushort.max);
+		enableMetaData();
+	}
+
+private:
+	void enableMetaData() {
+		assert(scSupportsMetaData, "size class not supports slab metadata!");
+
+		if (!metaDataEnabled) {
+			e.enableMetaData(sg.index);
+			metaDataEnabled = true;
+		}
+	}
+
+	void disableMetaData() {
+		assert(scSupportsMetaData, "size class not supports slab metadata!");
+
+		if (metaDataEnabled) {
+			e.disableMetaData(sg.index);
+			metaDataEnabled = false;
+		}
+	}
+
+	@property
+	T* ptrToAllocEnd(T)() {
+		return cast(T*) (sg.address + sg.size) - T.sizeof;
+	}
+
+	alias freeSpacePtr = ptrToAllocEnd!ushort;
+}
+
+unittest SlabAllocInfo {
+	import d.gc.tcache;
+	// When metadata is not supported by the size class:
+	foreach (size; [1, 6, 8, 20, 24, 35, 40, 50, 56]) {
+		auto s = threadCache.alloc(size, false);
+		auto pd = threadCache.getPageDescriptor(s);
```

Comment:
You don't need the thread cache to be involved here. The goal is to test the SlabAllocInfo, not the whole GC, so that you know that when this test break, you have a problem in SlabAllocInfo, not somewhere in the whole GC.

Allocate an extent on stack, have it manage some chunk of memory and go from there.